x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -143,12 +150,106 @@ public void execute(
         analyzedPlan(
             parse(request.query(), request.params()),
             executionInfo,
-            listener.delegateFailureAndWrap(
-                (next, analyzedPlan) -> executeOptimizedPlan(request, executionInfo, runPhase, optimizedPlan(analyzedPlan), next)
-            )
+            new LogicalPlanActionListener(request, executionInfo, runPhase, listener)
         );
     }
 
+    /**
+     * ActionListener that receives LogicalPlan or error from logical planning.
+     * Any Exception sent to onFailure stops processing, but not all are fatal (return a 4xx or 5xx), so
+     * the onFailure handler determines whether to return an empty successful result or a 4xx/5xx error.
+     */
+    class LogicalPlanActionListener implements ActionListener<LogicalPlan> {
+        private final EsqlQueryRequest request;
+        private final EsqlExecutionInfo executionInfo;
+        private final BiConsumer<PhysicalPlan, ActionListener<Result>> runPhase;
+        private final ActionListener<Result> listener;
+
+        LogicalPlanActionListener(
+            EsqlQueryRequest request,
+            EsqlExecutionInfo executionInfo,
+            BiConsumer<PhysicalPlan, ActionListener<Result>> runPhase,
+            ActionListener<Result> listener
+        ) {
+            this.request = request;
+            this.executionInfo = executionInfo;
+            this.runPhase = runPhase;
+            this.listener = listener;
+        }
+
+        @Override
+        public void onResponse(LogicalPlan analyzedPlan) {
+            executeOptimizedPlan(request, executionInfo, runPhase, optimizedPlan(analyzedPlan), listener);
+        }
+
+        /**
+         * Whether to return an empty result (HTTP status 200) for a CCS rather than a top level 4xx/5xx error.
+         *
+         * For cases where field-caps had no indices to search and the remotes were unavailable, we
```

Comment:
First comment - I was planning to handle the "missing indices" questions in "ticket 2" of the skip_unavailable work (this is ticket 1 dealing with just disconnected clusters).

That said, let's clarify here what behavior you are proposing:

1) If, at the end of cluster resolution (and checking for connectivity) and index resolution, there are no indices to query (either because the clusters are unavailable and/or the available clusters have no matching indices), then we should throw a VerificationException, rather than an empty "successful" result.

2) The above is true even if it is a remote-only query and wildcards were used for either the cluster names or the index names or both.

If I have that right - do you think there is any scenario where **at planning time** (before starting ComputeService execution) we should return an empty "successful" result?

Reply:
> 1. If, at the end of cluster resolution (and checking for connectivity) and index resolution, there are no indices to query (either because the clusters are unavailable and/or the available clusters have no matching indices), then we should throw a VerificationException, rather than an empty "successful" result.

No :-). My comment was specific to `from inexistent, remote:inexistent` and `from inexistent, remote:inexistent*`. This is a scenario that I missed to cover when we discussed about behavior of CCS queries. This involves the **local cluster** as well and I _think_ in this case it doesn't matter how CCS behaves. The fact that the **local cluster doesn't have an explicitly specified index** should fail no matter what. In the two tables from https://github.com/elastic/elasticsearch/issues/112886 (skip_un true/false) a better comparison query than the one in my previous review comment is 
> from remote:existent, inexistent

where the behavior is 

> failure coming from ES at query time complaining about "inexistent" from the "local" cluster

But, in our case the failure should be coming from the Verifier because the `field_caps` call returned no indices.

Another two use cases are `from inexistent*, remote:inexistent` and `from inexistent*, remote:inexistent*` and the principle for both these two cases and the previous two should be _if the local cluster is involved and the overall - remote and local - resolution of indices ends up with no indices this should be an error because the user specified a missing indices pattern for the local cluster; even if the user cannot control how the remote clusters behave, the user is in control of the index pattern for the local cluster_.

**skip_unavailable = true/false**

|index pattern                   | behavior                                                                                                                     |
|----------------------------|------------------------------------------------------------------------------------------|
| from inexistent*, remote:inexistent | failure     coming from post-`_field_caps` check, at the `Verifier` level |
| from inexistent*, remote:inexistent*  |  failure     coming from post-`_field_caps` check, at the `Verifier` level |

ESQL's current behavior where **no CCS** is used is to reject queries that have no indices after the index pattern is resolved:
`from inexistent*`
`from inexistent1*,inexistent2*`
`from inexistent`
`from inexistent1,inexistent2`
`from inexistent,inexistent1*`

I have updated the tables from https://github.com/elastic/elasticsearch/issues/112886